### ISSUE TITLE
Ability to programmatically attach to own JVM

### ISSUE THREAD

**Issue by robrez** [microsoft/ApplicationInsights-Java#2125]
# Problem

Would like parity with the dotnet cohorts who are able to add applicationinsights as a dependency and programatically add instrumentation to an application, [example](https://docs.microsoft.com/en-us/azure/azure-monitor/app/asp-net-core#enable-application-insights-server-side-telemetry-no-visual-studio)

For java, it seems like the only vended option is to use a javaagent at startup.  Would like the ability to do something like this:

```java
  public static void main(String[] args) throws IOException {
    ApplicationInsights.attach();
    SpringApplication.run(SpringBootApp.class, args);
  }
 ```

# Potential Solution

The snippet above is from _this_ codebase, but it's inside test packages

- https://github.com/microsoft/ApplicationInsights-Java/blob/main/test/smoke/testApps/SpringBootAttachInMain/src/main/java/com/microsoft/applicationinsights/smoketestapp/SpringBootApp.java
- https://github.com/microsoft/ApplicationInsights-Java/blob/main/test/smoke/testApps/SpringBootAttachInMain/src/main/java/com/microsoft/applicationinsights/smoketestapp/AttachInFirstLineOfMain.java

I don't mean to link to a competitor.. but this source is apache2-licensed and offers another example

- https://github.com/elastic/apm-agent-java/blob/HEAD/apm-agent-attach/src/main/java/co/elastic/apm/attach/ElasticApmAttacher.java#L57

**Comment by trask** [comment]
hi @robrez, you can do this today (as the test shows), so I'm not clear what your request is?

**Comment by robrez** [comment]
Thanks for the quick response.. was just hoping to be able to do this with something vended within the application insights jar. 
 I could make my own class that's similar to the one linked above, but worry I'd get some details wrong that would be necessary for production quality.  We're looking to add insights to a number of dotnet and java applications at my org;  In my opinion, this addition would be a valuable dx improvement.

**Comment by trask** [comment]
can you try using https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/runtime-attach? if it's missing anything I'm happy to add it over there, we are trying to keep the vended API surface to a minimum

**Comment by robrez** [comment]
Added this to my application and, yes it does seem to have worked.

```java
io.opentelemetry.contrib.attach.RuntimeAttach.attachJavaagentToCurrentJVM();
```

I need to educate myself on how. I'll be honest, I didn't expect that to work since the `applicationinsights-agent.jar` is... a different jar from [this one](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/runtime-attach/src/main/java/io/opentelemetry/contrib/attach/AgentFileLocator.java#L17)

I noticed that some of the other code snippets (above) are dong some extra-curriculars... creating temp files, copying the agent jar to a temp location, etc. I'm also wondering if those details matter?

I'll proceed via the runtime-attach strategy and file issues over there should anything come up.

**Comment by trask** [comment]
> I didn't expect that to work since the `applicationinsights-agent.jar` is... a different jar from [this one](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/runtime-attach/src/main/java/io/opentelemetry/contrib/attach/AgentFileLocator.java#L17)

The application insights agent is a distro of the opentelemetry javaagent, so that class `io.opentelemetry.javaagent.OpenTelemetryAgent` exists in our distro too 😄.

> I noticed that some of the other code snippets (above) are dong some extra-curriculars... creating temp files, copying the agent jar to a temp location, etc. I'm also wondering if those details matter?

IIRC that's needed if you are bundling the agent inside of a spring boot uber jar (it copies it the agent jar out of the spring boot uber jar), otherwise no need to do that.

> I'll proceed via the runtime-attach strategy and file issues over there should anything come up.

👍

**Comment by robrez** [comment]
I did end up needing to perform a lot of the "extra curricular" activity mentioned above (copying the agent jar) for some projects within my org because they are using spring-boot runnable jars.   For these applications, the opentel helper didn't quite work.

FWIW, those projects could exclude the insight agent from the uber jar and use some other technique for placing it on the classpath.

Please feel free to close this issue!  I would still love to see some lib to meet the need, but it seems like it would be a bit challenging to keep the opentel helper "pure".  I think you could probably change it to extract the agent jar to a temp location, but moving `applicationinsights.json` there wouldn't quite fit... although that detail could be sorted out some other way

**Comment by trask** [comment]
hey @robrez, I would love to get these changes upstream to the otel helper where possible. if you could either send PR to the otel contrib repo, or point me at what you wrote and I can extract common stuff into the otel contrib, I think that would be really useful

**Comment by trask** [comment]
Re-opening, as we think something like this is a good idea:

> ```java
>   public static void main(String[] args) throws IOException {
>     ApplicationInsights.attach();
>     SpringApplication.run(SpringBootApp.class, args);
>   }
>  ```

and agree that

> this addition would be a valuable dx improvement

**Comment by trask** [comment]
Addressed by #2325
